### PR TITLE
12 device location

### PR DESCRIPTION
**Pull request by ArtemKlots-in6k**

### CODE REVIEW

File: client/src/app/devices/models/model.service.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import {Injectable} from '@angular/core';
 import {Http} from '@angular/http';
 
 import {Model} from "./model";
-import { MODELS } from './mock-models';
+import {MODELS} from './mock-models';
```
Comment (alexanderko) [inline client/src/app/devices/models/model.service.ts]: remove mock dependency

File: client/src/app/devices/input-auto-complete.component.ts [hunk 2]
```diff
@@ -116,4 +117,40 @@ export class InputAutoComplete {
...
         return (str+'').replace(/([\\\.\+\*\?\[\^\]\$\(\)\{\}\=\!\<\>\|\:])/g, "\\$1");
     }
+
+    moveByArrows(direction: string) {
+        if (direction == 'down') {
+            this.moveToNext(1, () => {
+                return _.head(this.filteredList);
+            });
+        }
+        else {
```
Comment (alexanderko) [inline client/src/app/devices/input-auto-complete.component.ts]: `} else {`

File: client/src/app/devices/device.ts [hunk 3]
```diff
@@ -1,13 +1,14 @@
...
+import {Item} from "./Item";
 
 export class Device {
-
     public id: string = '';
     public type: Type = new Type();
     public vendor: Vendor = new Vendor();
     public model: string = '';
     public serialNumber: string = '';
+    public location = [];
```
Comment (alexanderko) [inline client/src/app/devices/device.ts]: specify type if possible

File: client/src/app/devices/models/model.service.ts (hunk 1 above)
Comment (YaroslavP-in6k) [inline client/src/app/devices/models/model.service.ts]: done

File: client/src/app/devices/input-auto-complete.component.ts (hunk 2 above)
Comment (YaroslavP-in6k) [inline client/src/app/devices/input-auto-complete.component.ts]: done

File: client/src/app/devices/device.ts (hunk 3 above)
Comment (YaroslavP-in6k) [inline client/src/app/devices/device.ts]: done

File: client/src/app/devices/areas/area-node.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+export class AreaNode {
```
Comment (alexanderko) [inline client/src/app/devices/areas/area-node.ts]: just `Area`

File: client/src/app/devices/areas/area-node.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+    id: string;
+    name: string;
+    sublocations: AreaNode[];
```
Comment (alexanderko) [inline client/src/app/devices/areas/area-node.ts]: rename `sublocations` to `areas`

File: api/src/main/java/com/in6k/vault13/dao/AreaTraversal.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+import com.arangodb.ArangoException;
+import com.in6k.vault13.entity.Area;
+import com.in6k.vault13.entity.AreaNode;
+
+import java.util.List;
+
+public interface AreaTraversal {
+
+    List<AreaNode> getSublocationsById(String locationId) throws ArangoException;
```
Comment (alexanderko) [inline api/src/main/java/com/in6k/vault13/dao/AreaTraversal.java]: rename to `getAreasByParentId`

File: api/src/main/java/com/in6k/vault13/dao/AreasDao.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package com.in6k.vault13.dao;
+
+import com.arangodb.ArangoException;
+import com.in6k.vault13.entity.Area;
+
+import java.util.List;
+
+public interface AreasDao {
+
+    List<Area> getAllAreasOfFirstOrder() throws Exception;
```
Comment (alexanderko) [inline api/src/main/java/com/in6k/vault13/dao/AreasDao.java]: remove unused method

File: api/src/main/java/com/in6k/vault13/dao/AreasDao.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+    Area getById(String id) throws ArangoException;
+
+    Area add(Area area);
+
+    void linkAreaWithSublocation(Area location, Area sublocation);
```
Comment (alexanderko) [inline api/src/main/java/com/in6k/vault13/dao/AreasDao.java]: rename to `linkAreas(Area broaderArea, Area narrowerArea)`

File: api/src/main/java/com/in6k/vault13/dao/impl/AreaTraversalImpl.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        do {
+            String query = areaTraversalQueryBuilder.getLocationPosition(areaNode.getLocatedAt());
+            areaNode = DaoUtils.getUniqueResult(
+                    databaseConnector.getArangoDriver(),
+                    query,
+                    AreaNode.class
+            );
+
+            result.add(new Area(areaNode.getId(), areaNode.getName()));
+        } while (!areaNode.getLocatedAt().equals(ArangoCollections.AREAS.getDocumentId("root")));
```
Comment (alexanderko) [inline api/src/main/java/com/in6k/vault13/dao/impl/AreaTraversalImpl.java]: extract condition into method with appropriate name

File: api/src/main/java/com/in6k/vault13/dao/impl/AreaTraversalImpl.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        return invertAreaOrder(result);
+    }
+
+    private List<Area> invertAreaOrder(List<Area> list) {
+        List<Area> result = new ArrayList<>();
+        for (int i = list.size() - 1; i >= 0; i--) {
```
Comment (alexanderko) [inline api/src/main/java/com/in6k/vault13/dao/impl/AreaTraversalImpl.java]: use algorithm from collections framework

File: client/src/app/devices/areas/area-node.ts (hunk 4 above)
Comment (YaroslavP-in6k) [inline client/src/app/devices/areas/area-node.ts]: done

File: api/src/main/java/com/in6k/vault13/dao/AreasDao.java (hunk 8 above)
Comment (ArtemKlots-in6k) [inline api/src/main/java/com/in6k/vault13/dao/AreasDao.java]: Done

File: api/src/main/java/com/in6k/vault13/dao/impl/DevicesDaoImpl.java [hunk 11]
```diff
@@ -120,7 +126,7 @@ public void linkVendorWithDevice(Device device, Vendor vendor) {
         try {
             BaseDocument vendorToAdd = new BaseDocument();
             vendorToAdd.addAttribute("_from", ArangoCollections.DEVICES.getDocumentId(device.getId()));
-            vendorToAdd.addAttribute("_to", ArangoCollections.VENDORS.getDocumentId(vendor.getId()));
+                vendorToAdd.addAttribute("_to", ArangoCollections.VENDORS.getDocumentId(vendor.getId()));
```
Comment (alexanderko) [inline api/src/main/java/com/in6k/vault13/dao/impl/DevicesDaoImpl.java]: remove extra space

File: api/src/main/java/com/in6k/vault13/dao/AreaTraversal.java (hunk 6 above)
Comment (ArtemKlots-in6k) [inline api/src/main/java/com/in6k/vault13/dao/AreaTraversal.java]: Done

File: api/src/main/java/com/in6k/vault13/dao/AreasDao.java (hunk 7 above)
Comment (ArtemKlots-in6k) [inline api/src/main/java/com/in6k/vault13/dao/AreasDao.java]: Done

File: api/src/main/java/com/in6k/vault13/dao/impl/DevicesDaoImpl.java (hunk 11 above)
Comment (YaroslavP-in6k) [inline api/src/main/java/com/in6k/vault13/dao/impl/DevicesDaoImpl.java]: done

File: api/src/main/java/com/in6k/vault13/dao/impl/AreaTraversalImpl.java (hunk 9 above)
Comment (YaroslavP-in6k) [inline api/src/main/java/com/in6k/vault13/dao/impl/AreaTraversalImpl.java]: done

**Pull request merged**